pkg/controllers/multicluster/cloudmap_controller.go

```diff
@@ -184,23 +205,24 @@ func (r *CloudMapReconciler) createAndGetServiceImport(ctx context.Context, name
 	return r.getServiceImport(ctx, namespace, name)
 }
 
-func (r *CloudMapReconciler) getDerivedService(ctx context.Context, namespace string, name string) (*v1.Service, error) {
+func (r *CloudMapReconciler) getDerivedService(ctx context.Context, namespace string, name string, clusterId string) (*v1.Service, error) {
+	derivedName := DerivedName(namespace, name, clusterId)
 	existingService := &v1.Service{}
-	err := r.Client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, existingService)
+	err := r.Client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: derivedName}, existingService)
 	return existingService, err
 }
 
-func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svcImport *multiclusterv1alpha1.ServiceImport, svcPorts []*model.Port) (*v1.Service, error) {
-	toCreate := CreateDerivedServiceStruct(svcImport, svcPorts)
+func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svcImport *multiclusterv1alpha1.ServiceImport, clusterId string, svcPorts []*model.Port) (*v1.Service, error) {
+	toCreate := CreateDerivedServiceStruct(svcImport, svcPorts, clusterId)
 	if err := r.Client.Create(ctx, toCreate); err != nil {
 		return nil, err
 	}
 	r.Log.Info("created derived Service", "namespace", toCreate.Namespace, "name", toCreate.Name)
 
-	return r.getDerivedService(ctx, toCreate.Namespace, svcImport.Annotations[DerivedServiceAnnotation])
+	return r.getDerivedService(ctx, svcImport.Namespace, svcImport.Name, clusterId)
 }
 
-func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *multiclusterv1alpha1.ServiceImport, desiredEndpoints []*model.Endpoint, svc *v1.Service) error {
+func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *multiclusterv1alpha1.ServiceImport, desiredEndpoints []*model.Endpoint, svc *v1.Service, clusterId string) error {
 	existingSlicesList := discovery.EndpointSliceList{}
 	if err := r.Client.List(ctx, &existingSlicesList,
 		client.InNamespace(svc.Namespace), client.MatchingLabels{discovery.LabelServiceName: svc.Name}); err != nil {
```

Comment:
Needs to also filter by label `multicluster.kubernetes.io/source-cluster`

Reply:
Don't believe that is necessary. The svc.Name is the derivedService name. Therefore, filtering EndpointSlices based on the derived service name is sufficient to get its endpoint slices.

Derived Service Name will be unique as the name is built with hash(ns, svc-name, clusterId). If clusterId is different, then derived name will be unique and we do not need to filter by source cluster. 

(Unless we are worried about hash conflicts?)